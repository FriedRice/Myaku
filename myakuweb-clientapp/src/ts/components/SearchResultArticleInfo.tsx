/** @module Search result article info list component */

import { ArticleSearchResult } from '../types';
import React from 'react';

import {
    getDaysBetween,
    humanizeDate,
} from '../utils';

interface SearchResultArticleInfoProps {
    searchResult: ArticleSearchResult;
}
type Props = SearchResultArticleInfoProps;


function makeTimeElement(date: Date): React.ReactElement {
    return (
        <time dateTime={date.toISOString()}>
            {humanizeDate(date)}
        </time>
    );
}

/**
 * Determine if last updated datetime should be displayed for the search
 * result.
 *
 * The last updated datetime is not worth displaying if it's close to the
 * publication date.
 *
 * Additionally, if the publication date is older, the amount of time between
 * publication and the update must be greater in order for the update to be
 * worth displaying.
 *
 * @param searchResult - Search result to determine if its last updated
 * datetime should be displayed.
 *
 * @returns True if the last updated datetime should be displayed, or False if
 * it shouldn't be displayed.
 */
function shouldDisplayLastUpdatedDate(
    searchResult: ArticleSearchResult
): boolean {
    if (searchResult.lastUpdatedDatetime === null) {
        return false;
    }

    var now = new Date();
    var daysSincePublish = getDaysBetween(
        now,
        searchResult.publicationDatetime
    );
    var daysSinceUpdate = getDaysBetween(
        now,
        searchResult.lastUpdatedDatetime
    );
    var daysUpdatedAfter = getDaysBetween(
        searchResult.lastUpdatedDatetime,
        searchResult.publicationDatetime
    );

    if (daysUpdatedAfter < 1 || daysSincePublish <= 7) {
        return false;
    }

    if (
        daysSinceUpdate <= 7
        || daysSincePublish < 180 && daysUpdatedAfter > 30
        || daysSincePublish < 365 && daysUpdatedAfter > 90
        || daysSincePublish < 365 * 2 && daysUpdatedAfter > 180
        || daysUpdatedAfter > 365
    ) {
        return true;
    }

    return false;
}

const SearchResultArticleInfo: React.FC<Props> = function(props) {
    var lastUpdatedDatetimeLi: React.ReactElement | null = null;
    if (shouldDisplayLastUpdatedDate(props.searchResult)) {
        lastUpdatedDatetimeLi = (
            <li key='updated-datetime' className='updated-datetime'>
                {'Updated '}
                {makeTimeElement(
                    props.searchResult.lastUpdatedDatetime as Date
                )}
            </li>
        );
    }

    return (
        <ul className='article-info'>
            <li key='source-name'>
                {props.searchResult.sourceName}
            </li>
            <li key='publication-datetime'>
                {makeTimeElement(props.searchResult.publicationDatetime)}
            </li>
            {lastUpdatedDatetimeLi}
            <li key='instance-count'>
                {props.searchResult.instanceCount.toLocaleString('en-US')}
                {' instances'}
            </li>
        </ul>
    );
};

export default SearchResultArticleInfo;
