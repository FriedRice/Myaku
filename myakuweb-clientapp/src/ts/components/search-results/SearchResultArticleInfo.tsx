/**
 * SearchResultArticleInfo component module. See [[SearchResultArticleInfo]].
 */

import { ArticleSearchResult } from 'ts/types/types';
import React from 'react';

import {
    getDaysBetween,
    humanizeDate,
} from 'ts/app/utils';

/** Props for the [[SearchResultArticleInfo]] component. */
interface SearchResultArticleInfoProps {
    searchResult: ArticleSearchResult;
}
type Props = SearchResultArticleInfoProps;


/**
 * Make a time element for the given date.
 *
 * @param date - Date to make a time element for.
 *
 * @returns The time element made for the date.
 */
function makeTimeElement(date: Date): React.ReactElement {
    return (
        <time dateTime={date.toISOString()}>
            {humanizeDate(date)}
        </time>
    );
}

/**
 * Determine if last updated datetime should be displayed for the search
 * result.
 *
 * The last updated datetime is not worth displaying if it's close to the
 * publication date.
 *
 * Additionally, if the publication date is older, the amount of time between
 * publication and the update must be greater in order for the update to be
 * worth displaying.
 *
 * @param searchResult - Search result to determine if its last updated
 * datetime should be displayed.
 *
 * @returns True if the last updated datetime should be displayed, or false if
 * it shouldn't be displayed.
 */
function shouldDisplayLastUpdatedDate(
    searchResult: ArticleSearchResult
): boolean {
    if (searchResult.lastUpdatedDatetime === null) {
        return false;
    }

    var now = new Date();
    var daysSincePublish = getDaysBetween(
        now,
        searchResult.publicationDatetime
    );
    var daysSinceUpdate = getDaysBetween(
        now,
        searchResult.lastUpdatedDatetime
    );
    var daysUpdatedAfter = getDaysBetween(
        searchResult.lastUpdatedDatetime,
        searchResult.publicationDatetime
    );

    if (daysUpdatedAfter < 1 || daysSincePublish <= 7) {
        return false;
    }

    if (
        daysSinceUpdate <= 7
        || daysSincePublish < 180 && daysUpdatedAfter > 30
        || daysSincePublish < 365 && daysUpdatedAfter > 90
        || daysSincePublish < 365 * 2 && daysUpdatedAfter > 180
        || daysUpdatedAfter > 365
    ) {
        return true;
    }

    return false;
}

/**
 * Search result article info list component.
 *
 * @param props - See [[SearchResultArticleInfoProps]].
 */
const SearchResultArticleInfo: React.FC<Props> = function(props) {
    var lastUpdatedDatetimeLi: React.ReactElement | null = null;
    if (shouldDisplayLastUpdatedDate(props.searchResult)) {
        lastUpdatedDatetimeLi = (
            <li key='updated-datetime' className='updated-datetime'>
                {'Updated '}
                {makeTimeElement(
                    props.searchResult.lastUpdatedDatetime as Date
                )}
            </li>
        );
    }

    return (
        <ul className='article-info'>
            <li key='source-name'>
                {props.searchResult.sourceName}
            </li>
            <li key='publication-datetime'>
                {makeTimeElement(props.searchResult.publicationDatetime)}
            </li>
            {lastUpdatedDatetimeLi}
            <li key='instance-count'>
                {props.searchResult.instanceCount.toLocaleString('en-US')}
                {' instances'}
            </li>
        </ul>
    );
};

export default SearchResultArticleInfo;
