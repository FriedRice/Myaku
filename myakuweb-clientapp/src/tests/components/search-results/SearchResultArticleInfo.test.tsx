/**
 * Tests for the [[SearchResultArticleInfo]] component.
 */

import React from 'react';
import SearchResultArticleInfo from
    'ts/components/search-results/SearchResultArticleInfo';

import {
    BASELINE_SEARCH_RESULT_PAGE,
    getSearchResultPageDataClone,
} from 'tests/testData';
import {
    ShallowWrapper,
    render,
    shallow,
} from 'enzyme';

var searchResult = BASELINE_SEARCH_RESULT_PAGE.results[0];
beforeEach(function() {
    searchResult = getSearchResultPageDataClone().results[0];
});

function expectLastUpdatedTimeShowing(showing: boolean): void {
    const wrapper = shallow(
        <SearchResultArticleInfo
            searchResult={searchResult}
        />
    );

    const lastUpdatedTimeLi = wrapper.findWhere(
        function(node: ShallowWrapper) {
            return node.name() === 'li' && node.key() === 'updated-datetime';
        }
    );
    expect(lastUpdatedTimeLi).toHaveLength(Number(showing));
}

function getDateDaysBeforeNow(days: number): Date {
    const now = new Date();
    return new Date(now.setDate(now.getDate() - days));
}

function getDateDaysBeforeDate(date: Date, days: number): Date {
    const clone = new Date(date.getTime());
    return new Date(clone.setDate(clone.getDate() - days));
}

function getDateDaysAfterDate(date: Date, days: number): Date {
    const clone = new Date(date.getTime());
    return new Date(clone.setDate(clone.getDate() + days));
}

function getDateMinutesAfterDate(date: Date, minutes: number): Date {
    const clone = new Date(date.getTime());
    return new Date(clone.setMinutes(clone.getMinutes() + minutes));
}


describe('<SearchResultArticleInfo /> render', function() {
    it('renders correctly with last updated time', function() {
        const wrapper = shallow(
            <SearchResultArticleInfo
                searchResult={searchResult}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly without last updated time', function() {
        searchResult.lastUpdatedDatetime = null;

        const wrapper = shallow(
            <SearchResultArticleInfo
                searchResult={searchResult}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('uses int comma for instance count', function() {
        searchResult.instanceCount = 1222333;

        const wrapper = render(
            <SearchResultArticleInfo
                searchResult={searchResult}
            />
        );
        expect(wrapper.text()).toStrictEqual(
            expect.stringContaining('1,222,333 instances')
        );
    });
});

describe('<SearchResultArticleInfo /> last updated time', function() {
    it('does not display if same day as publication', function() {
        searchResult.lastUpdatedDatetime = getDateMinutesAfterDate(
            searchResult.publicationDatetime, 1
        );
        expectLastUpdatedTimeShowing(false);
    });

    it('does not display if publication within week of now', function() {
        searchResult.publicationDatetime = getDateDaysBeforeNow(6);
        searchResult.lastUpdatedDatetime = getDateDaysAfterDate(
            searchResult.publicationDatetime, 2
        );
        expectLastUpdatedTimeShowing(false);
    });

    it('does display if update within week of now', function() {
        searchResult.lastUpdatedDatetime = getDateDaysBeforeNow(6);
        searchResult.publicationDatetime = getDateDaysBeforeDate(
            searchResult.lastUpdatedDatetime, 2
        );
        expectLastUpdatedTimeShowing(true);
    });

    it('does display if publish <= 180d and updated after 30d', function() {
        searchResult.publicationDatetime = getDateDaysBeforeNow(150);
        searchResult.lastUpdatedDatetime = getDateDaysAfterDate(
            searchResult.publicationDatetime, 30
        );
        expectLastUpdatedTimeShowing(false);

        searchResult.publicationDatetime = getDateDaysBeforeNow(150);
        searchResult.lastUpdatedDatetime = getDateDaysAfterDate(
            searchResult.publicationDatetime, 31
        );
        expectLastUpdatedTimeShowing(true);
    });

    it('does display if publish <= 365d and updated after 90d', function() {
        searchResult.publicationDatetime = getDateDaysBeforeNow(300);
        searchResult.lastUpdatedDatetime = getDateDaysAfterDate(
            searchResult.publicationDatetime, 90
        );
        expectLastUpdatedTimeShowing(false);

        searchResult.publicationDatetime = getDateDaysBeforeNow(300);
        searchResult.lastUpdatedDatetime = getDateDaysAfterDate(
            searchResult.publicationDatetime, 91
        );
        expectLastUpdatedTimeShowing(true);
    });

    it('does display if publish <= 730d and updated after 180d', function() {
        searchResult.publicationDatetime = getDateDaysBeforeNow(700);
        searchResult.lastUpdatedDatetime = getDateDaysAfterDate(
            searchResult.publicationDatetime, 180
        );
        expectLastUpdatedTimeShowing(false);

        searchResult.publicationDatetime = getDateDaysBeforeNow(700);
        searchResult.lastUpdatedDatetime = getDateDaysAfterDate(
            searchResult.publicationDatetime, 181
        );
        expectLastUpdatedTimeShowing(true);
    });

    it('does display if updated after 365d', function() {
        searchResult.publicationDatetime = getDateDaysBeforeNow(731);
        searchResult.lastUpdatedDatetime = getDateDaysAfterDate(
            searchResult.publicationDatetime, 365
        );
        expectLastUpdatedTimeShowing(false);

        searchResult.publicationDatetime = getDateDaysBeforeNow(731);
        searchResult.lastUpdatedDatetime = getDateDaysAfterDate(
            searchResult.publicationDatetime, 366
        );
        expectLastUpdatedTimeShowing(true);

        searchResult.publicationDatetime = getDateDaysBeforeNow(2000);
        searchResult.lastUpdatedDatetime = getDateDaysAfterDate(
            searchResult.publicationDatetime, 366
        );
        expectLastUpdatedTimeShowing(true);
    });
});
