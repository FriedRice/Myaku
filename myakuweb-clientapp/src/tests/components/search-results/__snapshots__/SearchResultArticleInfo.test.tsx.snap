// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResultArticleInfo /> render renders correctly with last updated time 1`] = `
<ul
  className="article-info"
>
  <li
    key="source-name"
  >
    Source 1
  </li>
  <li
    key="publication-datetime"
  >
    <time
      dateTime="2012-05-28T16:10:16.000Z"
    >
      May 28th, 2012
    </time>
  </li>
  <li
    className="updated-datetime"
    key="updated-datetime"
  >
    Updated 
    <time
      dateTime="2019-12-02T05:05:05.000Z"
    >
      Dec 2nd, 2019
    </time>
  </li>
  <li
    key="instance-count"
  >
    6
     instances
  </li>
</ul>
`;

exports[`<SearchResultArticleInfo /> render renders correctly without last updated time 1`] = `
<ul
  className="article-info"
>
  <li
    key="source-name"
  >
    Source 1
  </li>
  <li
    key="publication-datetime"
  >
    <time
      dateTime="2012-05-28T16:10:16.000Z"
    >
      May 28th, 2012
    </time>
  </li>
  <li
    key="instance-count"
  >
    6
     instances
  </li>
</ul>
`;
